### PR TITLE
feat(auth): UX improvements for expired connections

### PR DESCRIPTION
**Pull request by JadenSimon**
## Problem
* It's not always clear how to re-authenticate expired connections, especially when using multiple at the same time
* The "Switch Connections" menu doesn't display much info about the status of the connection

## Solution
* Update the "Switch Connections" picker to:
  * Add a description to expired/invalid connections
  * Show which tools (if any) a connection is attached to
  * Re-authenticate expired connections instead of switching to them
* Revamp the status bar to account for multiple connection scenarios

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/shared/telemetry/spans.ts [hunk 1]
```diff
@@ -162,6 +163,7 @@ export class TelemetrySpan {
 interface TelemetryContext {
     readonly spans: TelemetrySpan[]
     readonly activeSpan?: TelemetrySpan
+    readonly attributes?: Partial<MetricShapes[MetricName]>
```
Comment (JadenSimon) [inline src/shared/telemetry/spans.ts:166]: Something I've been meaning to add. I came across a use-case here so I just added it now.

File: src/credentials/wizards/connections.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import { telemetry } from '../../shared/telemetry/telemetry'
+import { createRefreshButton, createExitButton } from '../../shared/ui/buttons'
+import { createQuickPick, DataQuickPickItem } from '../../shared/ui/pickerPrompter'
+import { Auth, Connection, isIamConnection, reauthCommand } from '../auth'
+import { getDependentAuths } from '../secondaryAuth'
+
+export const getConnectionIcon = (conn: Connection) =>
+    conn.type === 'sso' ? getIcon('vscode-account') : getIcon('vscode-key')
+
+export function createConnectionPrompter(auth: Auth, type?: 'iam' | 'sso') {
```
Comment (JadenSimon) [inline src/credentials/wizards/connections.ts:19]: Had to split this out due to potential circular dependency issues. `secondaryAuth.ts` could possibly be consolidated into `auth.ts` to simplify things.

File: src/credentials/wizards/connections.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    const prompter = createQuickPick(loadItems(), {
+        placeholder,
+        title: localize('aws.auth.promptConnection.title', 'Switch Connection'),
+        buttons: [refreshButton, createExitButton()],
+    })
+
+    return prompter
+
+    async function loadItems(): Promise<DataQuickPickItem<Connection | 'addNewConnection' | 'editCredentials'>[]> {
```
Comment (JadenSimon) [inline src/credentials/wizards/connections.ts:36]: Placing function declarations _after_ the return statement of the outer scope is common in some codebases. Not sure if I like it or not though.

File: src/test/credentials/auth.test.ts [hunk 4]
```diff
@@ -14,7 +22,7 @@ import { ToolkitError } from '../../shared/errors'
 import { FakeMemento } from '../fakeExtensionContext'
 import { assertTreeItem } from '../shared/treeview/testUtil'
 import { getTestWindow } from '../shared/vscode/window'
-import { captureEvent } from '../testUtil'
+import { assertTelemetry, captureEvent, captureEventOnce } from '../testUtil'
 import { stub } from '../utilities/stubber'
 
 function createSsoProfile(props?: Partial<Omit<SsoProfile, 'type'>>): SsoProfile {
```
Comment (JadenSimon) [inline src/test/credentials/auth.test.ts:28]: TODO: extract some of these functions out so they can be used elsewhere.

File: src/credentials/statusBarItem.ts [hunk 5]
```diff
@@ -49,26 +52,40 @@ function handleDevSettings(devSettings: DevSettings, statusBarItem: vscode.Statu
...
 }
 
-// Resolves when the status bar reaches its final state
-export async function updateCredentialsStatusBarItem(
-    statusBarItem: vscode.StatusBarItem,
-    credentialsId?: string,
-    color?: vscode.ThemeColor
-): Promise<void> {
-    const connectedMsg = localize(
+function updateItem(statusBarItem: vscode.StatusBarItem): void {
```
Comment (JadenSimon) [inline src/credentials/statusBarItem.ts:55]: I deleted the old test file as it really didn't do much. Need to extract some `Auth` test util functions before it's worth writing tests for this.

File: src/credentials/wizards/connections.ts (hunk 2 above)
Comment (JadenSimon) [reply src/credentials/wizards/connections.ts]: Nevermind. Splitting up the file isn't a huge amount of work but it does require moving around a lot of stuff which I'd rather do in a separate PR.

File: src/webviews/components/settingsPanel.vue [hunk 6]
```diff
@@ -141,6 +141,7 @@ input[type='checkbox'] ~ label .collapse-button {
     color: var(--vscode-foreground);
     transition: transform 0.5s;
     transform: rotate(180deg);
+    text-align: right;
```
Comment (JadenSimon) [inline src/webviews/components/settingsPanel.vue:144]: Little bug I noticed. Same with the icons/css changes.

**Review by justinmk3 — APPROVED** [review]

**Comment by justinmk3** [comment]
Is the attributes idea discussed more somewhere else?

**Comment by JadenSimon** [comment]
> Is the attributes idea discussed more somewhere else?

No, I'll have to write something up. The goal is to generalize what we already do for "common metadata" i.e. attaching extra data to all future actions in the current context.

**Pull request merged**